### DISCUSSION TITLE
How to type annotate in a Declarative Table nested inside another class

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#9555] [Usage Questions]
In the process of migrating my code from SQLAlchemy 1.4 to 2.0, I'm running into issues trying to follow the instructions outlined here https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html for using `mapped_column` and adding type annotations for everything. The problem seems to be caused by the fact that my Declarative Tables are nested inside an enclosing class.

Here's a simple example demonstrating the issue:
```
from sqlalchemy.orm import DeclarativeBase, relationship, Mapped, mapped_column
from sqlalchemy import select, ForeignKey
from typing import List

class SchemaV1:

    class Base(DeclarativeBase):
        pass
    
    class A(Base):
        __tablename__ = "a"
        
        id: Mapped[int] = mapped_column(primary_key=True)
        b: Mapped[List["B"]] = relationship()
        
    class B(Base):
        __tablename__ = "b"
        
        id: Mapped[int] = mapped_column(primary_key=True)
        a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
        
statement = select(SchemaV1.A).join(SchemaV1.A.b)
print(statement)
```

The above code produces a mypy error:
```
> b: Mapped[List["B"]] = relationship()
error: Name "B" is not defined  [name-defined]
```

So I tried changing that line to:
```
-     b: Mapped[List["B"]] = relationship()
+     b: Mapped[List["SchemaV1.B"]] = relationship()
```
Now mypy is happy, but SQLAlchemy is unhappy:
```
sqlalchemy.exc.InvalidRequestError: When initializing mapper Mapper[A(a)], expression 'SchemaV1.B' failed to locate a name ('SchemaV1.B'). If this is a class name, consider adding this relationship() to the <class '__main__.SchemaV1.A'> class after both dependent classes have been defined.
```

Is there any way to appease both SQLAlchemy and mypy when you have your Declarative Tables defined inside an enclosing class like this?

**Comment by CaselIT** [comment]
Hi

I don't think that pattern is tested at all.
I would do as in 1.4, passing the target in the `relationship()`. This allows you to satisfy both mypy and sqlalchemy

**Reply by dkbarn** [reply]
Should I file this as a bug? Or is it not a pattern that SQLAlchemy should ever be expected to support?

**Reply by zzzeek** [reply]
I dont think this is a pattern we can support, unfortunately

**Reply by CaselIT** [reply]
Reason is that `.` is already used for modules (I think), so it would be hard to reuse it to look into classes.

@zzzeek I've tried finding docs on the `.` usage in the relationship and I've not found much

**Reply by zzzeek** [reply]
the dot is going to be module separator, it's here: https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#late-evaluation-of-relationship-arguments

if there's a one liner that makes this work, then sure.   I just dont know what's involved
